### PR TITLE
UOE-9179: Cache OpenWrap's DB details

### PR DESCRIPTION
**Pull request by AvinashKapre**
# Description

Please add change description or link to ticket, docs, etc.

# Checklist:

- [x] PR commit list is unique (rebase/pull with the origin branch to keep master clean).
- [x] JIRA number is added in the PR title and the commit message.
- [x] Updated the `header-bidding` repo with appropiate commit id.
- [x] Documented the new changes.

For Prebid upgrade, refer: https://inside.pubmatic.com:8443/confluence/display/Products/Prebid-server+upgrade

### CODE REVIEW

File: modules/pubmatic/openwrap/cache/gocache/fsc_test.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+				cache: gocache.New(100, 100),
+				db:    mockDatabase,
+				cfg: config.Cache{
+					CacheDefaultExpiry: 1000,
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "Error In DB, Set Empty",
```
Comment (aadit-patil) [inline modules/pubmatic/openwrap/cache/gocache/fsc_test.go:134]: can we add one more case invalid / floating percentage given as input, will be neglected you can refer existing _"Valid rows returned,avoiding floating pcnt values"_

File: modules/pubmatic/openwrap/cache/gocache/vast_tags_test.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+				db:    mockDatabase,
+				cfg: config.Cache{
+					VASTTagCacheExpiry: 100000,
+				},
+			},
+			args: args{
+				pubID: 5890,
+			},
+			want: map[int]*models.VASTTag{
+				101: {ID: 101, PartnerID: 501, URL: "vast_tag_url_1", Duration: 15, Price: 2.0},
```
Comment (pm-saurabh-narkhede) [inline modules/pubmatic/openwrap/cache/gocache/vast_tags_test.go:180]: add tests for duplicate vastTags

**Review by pm-saurabh-narkhede — DISMISSED** [review]

File: modules/pubmatic/openwrap/cache/gocache/fsc_test.go (hunk 1 above)
Comment (AvinashKapre) [reply modules/pubmatic/openwrap/cache/gocache/fsc_test.go:134]: already handled error in the mock.

File: modules/pubmatic/openwrap/cache.go [hunk 3]
```diff
@@ -7,18 +7,19 @@ import (
 	"github.com/prebid/prebid-server/openrtb_ext"
 )
 
-func getTestModePartnerConfigMap(pubid, profileid, displayversion int, platform string) map[int]map[string]string {
+func getTestModePartnerConfigMap(platform string, timeout int64, displayVersion int) map[int]map[string]string {
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/cache.go:10]: move this function to `profiledata.go` and delete this file

**Review by pm-nilesh-chate — CHANGES_REQUESTED** [review]
added a few comments

File: modules/pubmatic/openwrap/cache/gocache/adunit_config_test.go [hunk 4]
```diff
@@ -0,0 +1,547 @@
...
+		request        *openrtb2.BidRequest
+		pubID          int
+		profileID      int
+		displayVersion int
+	}
+	type want struct {
+		adunitConfig    json.RawMessage
+		cacheKeyPresent bool
+	}
+	tests := []struct {
```
[the remark this answers is not in the document]
Comment (AvinashKapre) [reply modules/pubmatic/openwrap/cache/gocache/adunit_config_test.go:296]: Done

File: modules/pubmatic/openwrap/cache/gocache/adunit_config_test.go [hunk 5]
```diff
@@ -0,0 +1,547 @@
...
...egex":true,"config":{"default":{"floors":{"floormin":15,"data":{"currency":"USD","modelgroups":[{"currency":"USD","modelweight":40,"modelversion":"model 1 from adunit config slot level","schema":{"fields":["mediaType","size","domain"],"delimiter":"|"},"values":{"*|728x90|www.website.com":13,"banner|300x250|www.website.com":1,"banner|300x600|*":4,"banner|728x90|www.website.com":5},"default":5}]},"enforcement":{"enforcejs":true,"enforcepbs":true,"enforcerate":100},"enabled":true},"video":{"enabled":true,"config":{"mimes":null,"minduration":10,"maxduration":50,"skip":1,"skipmin":10,"skipafter":15,"battr":[6,7],"connectiontype":[2]}},"universalpixel":[{"id":123,"pixel":"pixle","pixeltype":"js","pos":"above","mediatype":"banner","partners":["pubmatic","appnexus"]}]},"div1":{"banner":{"enabled":true,"config":{"format":[{"w":200,"h":300},{"w":500,"h":800}]}},"video":{"enabled":true,"config":{"mimes":null,"minduration":10,"maxduration":50,"skip":1,"skipmin":10,"skipafter":15,"battr":[6,7],"connectiontype":[0,1,2,4]}}},"div2":{"video":{"enabled":true,"config":{"mimes":null,"minduration":10,"maxduration":50,"skip":1,"skipmin":10,"skipafter":15,"battr":[6,7],"connectiontype":[0,1,2,4]}}}}}`),
```
[the remark this answers is not in the document]
Comment (AvinashKapre) [reply modules/pubmatic/openwrap/cache/gocache/adunit_config_test.go:191]: Done

File: modules/pubmatic/openwrap/cache/gocache/adunit_config_test.go [hunk 6]
```diff
@@ -0,0 +1,547 @@
...
+			}
+			cacheKey := key(PubAdunitConfig, tt.args.pubID, tt.args.profileID, tt.args.displayVersion)
+			obj, found := c.Get(cacheKey)
+
+			if !tt.want.wantErr && !found {
+				t.Error("Adunit Config not found in cache for cache key", cacheKey)
+				return
+			}
+
+			if obj != nil {
```
[the remark this answers is not in the document]
Comment (AvinashKapre) [reply modules/pubmatic/openwrap/cache/gocache/adunit_config_test.go:262]: Done

File: modules/pubmatic/openwrap/cache.go (hunk 3 above)
Comment (AvinashKapre) [reply modules/pubmatic/openwrap/cache.go:10]: Done

File: modules/pubmatic/openwrap/cache/gocache/vast_tags.go [hunk 7]
```diff
@@ -5,14 +5,11 @@ import (
 )
 
 // PopulatePublisherVASTTags will put publisher level VAST Tag details into cache
-func (c *cache) populatePublisherVASTTags(pubid int) {
```
[the remark this answers is not in the document]
Comment (AvinashKapre) [reply modules/pubmatic/openwrap/cache/gocache/vast_tags.go:8]: If we remove check from caller then it always call `LockAndLoad` 
`if mapNameHash, ok := c.cache.Get(key(PubSlotNameHash, pubID)); !ok || mapNameHash == nil {
		c.LockAndLoad(pubLockKey, func() error {
			dbAccessed = true
			return c.populateCacheWithPubSlotNameHash(pubID)
		})
		//TODO: Add stat if error from the DB
	}`

File: modules/pubmatic/openwrap/cache/gocache/adunit_config_test.go [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+			if !tt.want.wantErr == !found {
+				t.Error("Adunit Config not found in cache for cache key", cacheKey)
+				return
+			}
+			if obj != nil {
+				adunitConfig := obj.(*adunitconfig.AdUnitConfig)
+				assert.Equal(t, tt.want.adunitConfig, adunitConfig, "Expected: %v but got %v", tt.want.adunitConfig, adunitConfig)
+			}
+		})
+	}
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/cache/gocache/adunit_config_test.go:268]:
```
err := c.populateCacheWithAdunitConfig(tt.args.pubID, tt.args.profileID, tt.args.displayVersion)
if assert.Equal(t, wantErr, err != nil) {
    assert.Equal(t, tt.want.adunitConfig, gotAdunitConfig)
}  else {
    assert.Equal(t, nil, gotAdunitConfig)    
}
```

File: modules/pubmatic/openwrap/cache/gocache/slot_mappings.go [hunk 9]
```diff
@@ -11,37 +11,40 @@ import (
...
 }
 
 // PopulateCacheWithWrapperSlotMappings will get the SlotMappings from database and put them in cache.
-func (c *cache) populateCacheWithWrapperSlotMappings(pubid int, partnerConfigMap map[int]map[string]string, profileId, displayVersion int) {
-	partnerSlotMappingMap, _ := c.db.GetWrapperSlotMappings(partnerConfigMap, profileId, displayVersion)
+func (c *cache) populateCacheWithWrapperSlotMappings(pubID int, partnerConfigMap map[int]map[string]string, profileID, displayVersion int) (err error) {
+	partnerSlotMappingMap, err := c.db.GetWrapperSlotMappings(partnerConfigMap, profileID, displayVersion)
+	if err != nil {
+		return
+	}
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/cache/gocache/slot_mappings.go:31]: remove this check to keep current behaviour as is

Comment (AvinashKapre) [reply modules/pubmatic/openwrap/cache/gocache/slot_mappings.go]: Done

**Review by pm-saurabh-narkhede — APPROVED** [review]

**Review by pm-aadit-patil — APPROVED** [review]

**Review by pm-nilesh-chate — APPROVED** [review]
